FWCore/Framework/src/EventProcessor.cc

```diff
@@ -1139,377 +1116,688 @@ namespace edm {
-      beginStreamsTransitionAsync<Traits>(WaitingTaskHolder(taskGroup_, &streamLoopWaitTask),
-                                          *schedule_,
-                                          preallocations_.numberOfStreams(),
-                                          transitionInfo,
-                                          serviceToken_,
-                                          subProcesses_);
-      streamLoopWaitTask.wait();
-    }
-    FDEBUG(1) << "\tstreamBeginRun " << run << "\n";
-    if (looper_) {
-      //looper_->doStreamBeginRun(schedule_->streamID(),runPrincipal, es);
-    }
-  }
-
-  void EventProcessor::endUnfinishedRun(ProcessHistoryID const& phid,
-                                        RunNumber_t run,
-                                        bool globalBeginSucceeded,
-                                        bool cleaningUpAfterException,
-                                        bool eventSetupForInstanceSucceeded) {
-    if (eventSetupForInstanceSucceeded) {
-      //If we skip empty runs, this would be called conditionally
-      endRun(phid, run, globalBeginSucceeded, cleaningUpAfterException);
-
-      if (globalBeginSucceeded) {
-        RunPrincipal& runPrincipal = principalCache_.runPrincipal(phid, run);
-        if (runPrincipal.shouldWriteRun() != RunPrincipal::kNo) {
-          FinalWaitingTask t{taskGroup_};
-          MergeableRunProductMetadata* mergeableRunProductMetadata = runPrincipal.mergeableRunProductMetadata();
-          mergeableRunProductMetadata->preWriteRun();
-          writeRunAsync(edm::WaitingTaskHolder{taskGroup_, &t}, phid, run, mergeableRunProductMetadata);
-          auto exceptn = t.waitNoThrow();
-          mergeableRunProductMetadata->postWriteRun();
-          if (exceptn) {
-            std::rethrow_exception(exceptn);
-          }
+    auto status = std::make_shared<RunProcessingStatus>(preallocations_.numberOfStreams(), iHolder);
+
+    chain::first([this, &status, &iSync](auto nextTask) {
+      espController_->runOrQueueEventSetupForInstanceAsync(iSync,
+                                                           nextTask,
+                                                           status->endIOVWaitingTasks(),
+                                                           status->eventSetupImpls(),
+                                                           queueWhichWaitsForIOVsToFinish_,
+                                                           actReg_.get(),
+                                                           forceESCacheClearOnNewRun_);
+    }) | chain::then([this, status, iSync](std::exception_ptr const* iException, auto nextTask) {
+      CMS_SA_ALLOW try {
+        if (iException) {
+          WaitingTaskHolder copyHolder(nextTask);
+          copyHolder.doneWaiting(*iException);
         }
+        ServiceRegistry::Operate operate(serviceToken_);
+        actReg_->postESSyncIOVSignal_.emit(iSync);
+
+        runQueue_->pushAndPause(
```

Comment:
I added the suggested comment.

One could add an extra 5 lines of code and also handle the exception before ever pushing to the runQueue. I thought about doing that and almost did. But I didn't because it would duplicate the code that runs when the runQueue task starts and it also follows the preexisting pattern in beginLumiAsync... One could reasonably go either way, not much difference ...